File: src/main/java/com/snowflake/snowpark_java/types/MapType.java
package com.snowflake.snowpark_java.types;

import java.util.Objects;

/**
 * Map data type. This maps to OBJECT data type in Snowflake.
 *
 * @since 0.9.0
 */
public class MapType extends DataType {
  private final DataType keyType;
  private final DataType valueType;
  private final String typeName;

  MapType(DataType keyType, DataType valueType) {
    this.keyType = keyType;
    this.valueType = valueType;
    this.typeName = "MapType[" + keyType.typeName() + ", " + valueType.typeName() + "]";
  }

  /**
   * Retrieves the data type of this MapType's key.
   *
   * @return A DataType object representing the key type of this MapType
   * @since 0.9.0
   */
  public DataType getKeyType() {
    return keyType;
  }

  /**
   * Retrieves the data type of this MapType's value.
   *
   * @return A DataType object representing the value type of this MapType
   * @since 0.9.0
   */
  public DataType getValueType() {
    return valueType;
  }

  /**
   * Generates a String value to represent this map type.
   *
   * @return A String value in the format of "MapType['key type', 'value type']"
   * @since 0.9.0
   */
  @Override
  public String toString() {
    return typeName;
  }

  /**
   * Verifies if a DataType equals to this one.
   *
   * @param other A DataType object
   * @return true if these data types are equivalent, false for otherwise.
   * @since 0.9.0
   */
  @Override
  public boolean equals(Object other) {
    if (other instanceof MapType) {
      MapType mt = (MapType) other;
      return keyType.equals(mt.keyType) && valueType.equals(mt.valueType);
    }
    return false;
  }

  /**
   * Calculates the hash code of this Map Type Object.
   *
   * @return An int number representing the hash code value
   * @since 0.9.0
   */
  @Override
  public int hashCode() {
    return Objects.hash(keyType, valueType, typeName);
  }
}

File: src/main/scala/com/snowflake/snowpark/types/MapType.scala
package com.snowflake.snowpark.types

/** Map data type. This maps to OBJECT data type in Snowflake.
  * @since 0.1.0
  */
case class MapType(keyType: DataType, valueType: DataType) extends DataType {
  override def toString: String = {
    s"MapType[${keyType.toString}, ${valueType.toString}]"
  }

  override private[snowpark] def schemaString =
    s"Map"
}

private[snowpark] class StructuredMapType(
    override val keyType: DataType,
    override val valueType: DataType,
    val isValueNullable: Boolean)
    extends MapType(keyType, valueType) {
  override def toString: String = {
    s"MapType[${keyType.toString}, ${valueType.toString} nullable = $isValueNullable]"
  }

  override private[snowpark] def schemaString =
    s"Map[${keyType.schemaString}, ${valueType.schemaString} nullable = $isValueNullable]"
}

private[snowpark] object StructuredMapType {
  def apply(keyType: DataType, valueType: DataType, isValueType: Boolean): StructuredMapType =
    new StructuredMapType(keyType, valueType, isValueType)
}
